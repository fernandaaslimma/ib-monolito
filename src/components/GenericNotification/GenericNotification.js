import React, { Component, Fragment } from "react";
import { Container } from "../../styles/grid";
import { redirect } from "../../utils/redirect";
import { MODAL_TYPES } from "../common/Modal/Modal";
import ErrorBoundary from "../ErrorBoundary/ErrorBoundary";
import NotificationModal from "./NotificationModal";

class GenericNotification extends Component {
  constructor(props) {
    super(props);
    this.state = {
      checked: false,
      isExiting: false,
      notificationState: { id: "" },
      termsApproved: false
    };
    this.changeState = this.changeState.bind(this);
    this.onClickClose = this.onClickClose.bind(this);
    this.handleClose = this.handleClose.bind(this);
    this.openModalApproveConsent = this.openModalApproveConsent.bind(this);
    this.navigatedTo = this.navigatedTo.bind(this);
  }

  async componentDidMount() {
    const { notificated, setNotificationStatus, notification } = this.props;

    let notificationInfo;
    if (notification && notification.length > 0) {
      notificationInfo = notification.find(noti => {
        if (
          noti.id &&
          !notificated[`${noti.id}`] &&
          noti.type === "navigateTo"
        ) {
          return noti;
        }
      });

      if (notificationInfo && notificationInfo.parameters) {
        this.setState({
          notificationState: notificationInfo
        });
        setTimeout(() => {
          this.openModalApproveConsent();
        }, 0);
      } else {
        setNotificationStatus(`${notificationInfo?.id}`);

        redirect("/home");
      }
    } else {
      setNotificationStatus(`${notification?.id}`);

      redirect("/home");
    }
  }

  componentDidUpdate(_, prevState) {
    if (prevState.isExiting !== this.state.isExiting) {
      this.openModalApproveConsent();
    }
  }

  navigatedTo(url) {
    const { notificationState } = this.state;
    const { setNotificationStatus, closeModal } = this.props;
    closeModal();
    setNotificationStatus(`${notificationState.id}`);
    redirect(url);
  }

  changeState(object) {
    this.setState(object);
  }

  handleClose() {
    const { notificationState } = this.state;
    const { setNotificationStatus, closeModal } = this.props;
    closeModal();
    setNotificationStatus(`${notificationState.id}`);
    redirect("/home");
  }

  onClickClose() {
    if (this.state.isExiting) {
      this.setState({ isExiting: false });
    } else {
      this.setState({ isExiting: true });
    }
  }

  openModalApproveConsent() {
    const { isExiting, notificationState } = this.state;
    this.props.openModal({
      type: MODAL_TYPES.PAGE_AS_MODAL,
      width: "auto",
      children: () => {
        return (
          <Fragment>
            <NotificationModal
              isExiting={isExiting}
              onClickClose={this.onClickClose}
              notification={notificationState}
              navigatedTo={this.navigatedTo}
              handleClose={this.handleClose}
            />
          </Fragment>
        );
      }
    });
  }
  render() {
    const { error } = this.props;
    return (
      <ErrorBoundary errorStatus={error} {...this.props}>
        <Container />
      </ErrorBoundary>
    );
  }
}
export default GenericNotification;
