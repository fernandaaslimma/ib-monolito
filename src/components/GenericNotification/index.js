import { connect } from "redux-zero/react";

import GenericNotification from "./GenericNotification";
import notificationActions from "../../actions/notification";
import modalActions from "../../actions/modal";
import { combineActions } from "redux-zero/utils";

export default connect(
  ({ notification, openModal, closeModal, notificated }) => ({
    notification,
    openModal,
    closeModal,
    notificated
  }),
  combineActions(modalActions, notificationActions)
)(GenericNotification);
