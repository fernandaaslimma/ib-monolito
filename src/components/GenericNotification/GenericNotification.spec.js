import React from "react";
import { shallow } from "enzyme";
import GenericNotification from "./GenericNotification";

jest.mock("../../utils/i18n", () => ({
  translate: id => id,
  getLanguage: () => "en-US",
  isZhCN: () => false,
  getDateFieldPlaceholderByLocale: () => "MM/DD/YYYY"
}));

jest.mock("../../utils/redirect");
const redirectMock = require("../../utils/redirect").redirect;

const props = {
  notification: [
    {
      description:
        "Você tem um consentimento de compartilhamento de dados pendente de aprovação",
      title: "Open Banking - Consentimento Pendente",
      type: "navigateTo",
      displayMethod: "PopUp",
      id: "11043ff8-2530-4e7b-9191-d31d2cbc8370",
      parameters: [
        {
          type: "buttonLabel",
          value: "Seguir para autorização"
        },
        {
          type: "url",
          value:
            "/open-banking/consent?intent_id=urn:bocombbm:bb7b05b5-95ab-4ea7-aaff-52d6c2f549a4"
        }
      ]
    }
  ],
  openModal(config) {
    config.children();
  },
  notificated: {},
  closeModal: jest.fn(),
  setNotificationStatus: jest.fn()
};

describe("test", () => {
  it("Should match snapshot", () => {
    expect(shallow(<GenericNotification {...props} />)).toMatchSnapshot();
  });

  //   it("Should match snapshot without parameters", () => {
  //     const newProps = {
  //       ...props,
  //       notification: [
  //         {
  //           title: "Alteração dos termos e condições de uso",
  //           description:
  //             "Realizamos algumas mudanças nos itens de  Confidencialidade das informações dos usuários e Garantias, responsabilidades e danos dos termos e condições de Uso.",
  //           displayMethod: "PopUp",
  //           type: "navigatedTo",
  //           id: "111"
  //         }
  //       ]
  //     };
  //     const component = shallow(<GenericNotification {...newProps} />);
  //     const spy = jest.spyOn(component.instance(), "openModalApproveConsent");

  //     expect(spy).toHaveBeenCalled();
  //     // expect(props.setNotificationStatus).toHaveBeenCalledWith("111");
  //     // expect(redirectMock).toHaveBeenCalledWith("/home");
  //   });

  it("Should use notification from state", () => {
    const newProps = {
      ...props,
      notification: []
    };
    const component = shallow(<GenericNotification {...newProps} />);

    expect(component).toMatchSnapshot();
  });

  //TODO: This test needs to be updated later
  it("Should test navigatedTo", () => {
    const component = shallow(<GenericNotification {...props} />);

    component.instance().navigatedTo("url");

    expect(props.closeModal).toHaveBeenCalled();
    expect(redirectMock).toHaveBeenCalledWith("url");
    expect(props.setNotificationStatus).toHaveBeenCalledWith(
      "11043ff8-2530-4e7b-9191-d31d2cbc8370"
    );
  });

  it("Should test changeState", () => {
    const component = shallow(<GenericNotification {...props} />);
    const spy = jest.spyOn(component.instance(), "setState");

    component.instance().changeState({ test: 111 });

    expect(spy).toHaveBeenCalledWith({ test: 111 });
  });

  it("Should test handleClose", () => {
    const component = shallow(<GenericNotification {...props} />);

    component.instance().handleClose();

    expect(props.closeModal).toHaveBeenCalled();
  });

  it("Should change isExiting to true", () => {
    const component = shallow(<GenericNotification {...props} />);
    const spy = jest.spyOn(component.instance(), "setState");

    component.setState({
      isExiting: false
    });

    component.instance().onClickClose();

    expect(spy).toHaveBeenCalledWith({ isExiting: true });

    expect(component).toMatchSnapshot();
  });

  it("Should change isExiting to false", () => {
    const component = shallow(<GenericNotification {...props} />);
    const spy = jest.spyOn(component.instance(), "setState");

    component.setState({
      isExiting: true
    });

    component.instance().onClickClose();

    expect(spy).toHaveBeenCalledWith({ isExiting: false });

    expect(component).toMatchSnapshot();
  });
});
