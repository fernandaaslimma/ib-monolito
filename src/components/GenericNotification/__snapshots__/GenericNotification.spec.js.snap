// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Should change isExiting to false 1`] = `
<ErrorBoundary
  closeModal={
    [MockFunction] {
      "calls": Array [
        Array [],
        Array [],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  notificated={Object {}}
  notification={
    Array [
      Object {
        "description": "Você tem um consentimento de compartilhamento de dados pendente de aprovação",
        "displayMethod": "PopUp",
        "id": "11043ff8-2530-4e7b-9191-d31d2cbc8370",
        "parameters": Array [
          Object {
            "type": "buttonLabel",
            "value": "Seguir para autorização",
          },
          Object {
            "type": "url",
            "value": "/open-banking/consent?intent_id=urn:bocombbm:bb7b05b5-95ab-4ea7-aaff-52d6c2f549a4",
          },
        ],
        "title": "Open Banking - Consentimento Pendente",
        "type": "navigateTo",
      },
    ]
  }
  openModal={[Function]}
  setNotificationStatus={
    [MockFunction] {
      "calls": Array [
        Array [
          "undefined",
        ],
        Array [
          "11043ff8-2530-4e7b-9191-d31d2cbc8370",
        ],
        Array [
          "11043ff8-2530-4e7b-9191-d31d2cbc8370",
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
        Object {
          "type": "return",
          "value": undefined,
        },
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
>
  <Container />
</ErrorBoundary>
`;

exports[`test Should change isExiting to true 1`] = `
<ErrorBoundary
  closeModal={
    [MockFunction] {
      "calls": Array [
        Array [],
        Array [],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  notificated={Object {}}
  notification={
    Array [
      Object {
        "description": "Você tem um consentimento de compartilhamento de dados pendente de aprovação",
        "displayMethod": "PopUp",
        "id": "11043ff8-2530-4e7b-9191-d31d2cbc8370",
        "parameters": Array [
          Object {
            "type": "buttonLabel",
            "value": "Seguir para autorização",
          },
          Object {
            "type": "url",
            "value": "/open-banking/consent?intent_id=urn:bocombbm:bb7b05b5-95ab-4ea7-aaff-52d6c2f549a4",
          },
        ],
        "title": "Open Banking - Consentimento Pendente",
        "type": "navigateTo",
      },
    ]
  }
  openModal={[Function]}
  setNotificationStatus={
    [MockFunction] {
      "calls": Array [
        Array [
          "undefined",
        ],
        Array [
          "11043ff8-2530-4e7b-9191-d31d2cbc8370",
        ],
        Array [
          "11043ff8-2530-4e7b-9191-d31d2cbc8370",
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
        Object {
          "type": "return",
          "value": undefined,
        },
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
>
  <Container />
</ErrorBoundary>
`;

exports[`test Should match snapshot 1`] = `
<ErrorBoundary
  closeModal={[MockFunction]}
  notificated={Object {}}
  notification={
    Array [
      Object {
        "description": "Você tem um consentimento de compartilhamento de dados pendente de aprovação",
        "displayMethod": "PopUp",
        "id": "11043ff8-2530-4e7b-9191-d31d2cbc8370",
        "parameters": Array [
          Object {
            "type": "buttonLabel",
            "value": "Seguir para autorização",
          },
          Object {
            "type": "url",
            "value": "/open-banking/consent?intent_id=urn:bocombbm:bb7b05b5-95ab-4ea7-aaff-52d6c2f549a4",
          },
        ],
        "title": "Open Banking - Consentimento Pendente",
        "type": "navigateTo",
      },
    ]
  }
  openModal={[Function]}
  setNotificationStatus={[MockFunction]}
>
  <Container />
</ErrorBoundary>
`;

exports[`test Should use notification from state 1`] = `
<ErrorBoundary
  closeModal={[MockFunction]}
  notificated={Object {}}
  notification={Array []}
  openModal={[Function]}
  setNotificationStatus={
    [MockFunction] {
      "calls": Array [
        Array [
          "undefined",
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
>
  <Container />
</ErrorBoundary>
`;
